src/platform_impl/linux/mod.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright 2022-2022 Tauri Programme within The Commons Conservancy
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-License-Identifier: MIT
+
+mod icon;
+mod menu;
+mod tray;
+
+pub(crate) use icon::PlatformIcon;
+use tray::Tray;
+
+use crate::{icon::Icon, TrayIconAttributes, TrayIconId};
+
+pub struct TrayIcon {
+    tray_handle: ksni::Handle<Tray>,
+}
+
+impl TrayIcon {
+    pub fn new(id: TrayIconId, attrs: TrayIconAttributes) -> crate::Result<Self> {
+        let icon = attrs.icon.map(|icon| icon.inner.into());
+
+        let title = title_or_pkg_name(attrs.title.unwrap_or_default());
+
+        let tooltip = attrs.tooltip.unwrap_or_default();
+
+        let menu = attrs
+            .menu
+            .as_ref()
+            .map(|menu| menu.items().into_iter().map(Into::into).collect())
+            .unwrap_or_default();
+
+        let tray_service = ksni::TrayService::new(Tray::new(id, icon, title, tooltip, menu));
+        let tray_handle = tray_service.handle();
+        tray_service.spawn();
+
+        Ok(Self { tray_handle })
+    }
+
+    pub fn set_icon(&mut self, icon: Option<Icon>) -> crate::Result<()> {
+        let icon = icon.map(|icon| icon.inner.into());
+
+        self.tray_handle.update(|tray| {
+            tray.set_icon(icon);
+        });
+
+        Ok(())
+    }
+
+    pub fn set_menu(&mut self, menu: Option<Box<dyn crate::menu::ContextMenu>>) {
```

Comment:
it looks like the only way to change the tray menu is by calling this method, which is a breaking change. 

Previously, changing the state on the submenu, whether adding/removing items or changing the title/icon/check/enabled state of the item would also be reflected on the tray menu without overriding the menu again.

Reply:
Yes, you have to call set_menu explicitly.

Reply:
this is kind of a huge blocker, and we won't be able to merge as it will break tauri usage.

Reply:
I wonder if we can provide a hook in muda that fires some event when a context menu changes its items.

Reply:
The problem is, that the `ksni::TrayService` runs in a separate thread and the `muda::MenuItemKind` contains a inner value of type `Rc<RefCell<MenuChild>>`. So we can't send the menu items directly to the tray service thread. Ideally we would build the menu out of the latest version of the menu items whenever its displayed.
As far as I can see, a hook would need to be passed to the `muda::MenuItem` constructor, which would make the API very ugly.

I don't see a good solution here. So how about we add another field to the `muda::MenuItemKind` structs. The type would be `Arc<ArcSwap<MenuItem>>` with `MenuItem` being the enum that I added to `tray-icon`, which I would move to `muda`. Whenever the menu is changed, this value then needs to be updated as well. It also means that the data is duplicated. This is not great, but it should work relatively easily.

Reply:
Let's experiment with that idea, maybe even put it behind a `ksni` feature flag in `muda`

Reply:
It didn't work as I hoped. `ksni::Handle::update` must be called in order to update the menu. So I added a channel and a thread to do the update. Please have a look and tell me what you think.

Reply:
apologies for the delay, I just gave it a test and I am okay with the direction you're heading, I left a few comments here and at muda and once these are resolved I think we can move forward.

Reply:
Sounds good!